### PR TITLE
Adding Realm.distinct() method.

### LINKED ISSUE #960: DISTINCT
In my usage, there are various times where I only want one out of a set of objects that share a value of a field - much like the 'DISTINCT' clause in SQL. For exampe, with the items

name | logInTime
-----|-------
Bill | 1426270572
Mark | 1426272004
Mark | 1426273505
Mark | 1426274666
Sally | 1426268301
Sally | 1426271123

in the table 'User', `realm.where(User.class).distinct("name").findAll()` would return 3 items (one for 'Mark', one for 'Bill', and one for 'Sally'). With it like this, which 'Mark' and which 'Sally' is undefined - presumably the first one Realm finds.

A potential advanced aspect of this would be to select which of the rows to return. `realm.where(User.class).distinct("name", "logInTime", false).findAll()` would return `(Bill, 1426270572), (Mark, 1426274666), (Sally,1426271123)`, or only having the first discovered in an sorted RealmResult appear.

### PR DESCRIPTION
**Pull request by kneth**
Issue #960 requests a `distinct()` method. As Realm's underlying storage engine added support for the method recently, it is time to add such a method to Realm Java.

Closes #960 

@realm/java

### CODE REVIEW

File: realm-jni/src/io_realm_internal_table.cpp [hunk 1]
```diff
@@ -1315,17 +1315,24 @@ JNIEXPORT jlong JNICALL Java_io_realm_internal_Table_nativeGetDistinctView(
...
+        case type_Int:
+        case type_DateTime:
+        case type_String:
+            try {
+                TableView* pTableView = new TableView( pTable->get_distinct_view(S(columnIndex)) );
+                return reinterpret_cast<jlong>(pTableView);
+            } CATCH_STD()
+            break;
+        default:
+            ThrowException(env, IllegalArgument, "Invalid type - only bool, integer, datatime, string are supported.");
```
Comment (emanuelez) [inline realm-jni/src/io_realm_internal_table.cpp]: I think you should list the Java types here, in order not to confuse the users. Or maybe "float, Float, double and Double are not supported"

File: realm/src/main/java/io/realm/Realm.java [hunk 2]
```diff
@@ -946,6 +946,30 @@ boolean contains(Class<? extends RealmObject> clazz) {
     }
 
     /**
+     * Return a distinct set of objects of a specific class. If no objects exist, the
```
Comment (cmelchior) [inline realm/src/main/java/io/realm/Realm.java]: Also describe the policy for which element is returned for the distinct set (first, last, undefined)?

**Comment by cmelchior** [comment]
Great to get this in. I don't agree it completely resolves #960 though as `DISTINCT` on query results don't work with this, so either edit #960 to reflect that or create a new issue for tracking `distinct()` on `RealmResults`

File: realm/src/androidTest/java/io/realm/RealmTest.java [hunk 3]
```diff
@@ -2058,4 +2058,47 @@ private void awaitOrFail(CountDownLatch latch) {
...
+
+        RealmResults<AnnotationIndexTypes> distinctBool = testRealm.distinct(AnnotationIndexTypes.class, "indexBoolean");
+        assertEquals(2, distinctBool.size());
+
+        for (String fieldName : new String[]{"Long", "Date", "String"}) {
+            RealmResults<AnnotationIndexTypes> distinct = testRealm.distinct(AnnotationIndexTypes.class, "index" + fieldName);
+            assertEquals("index" + fieldName, numberOfBlocks, distinct.size());
+        }
+
+        for (String fieldName : new String[]{"Boolean", "Long", "Date", "String"}) {
```
Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: This should be a separate unit test.

File: realm/src/androidTest/java/io/realm/RealmTest.java [hunk 4]
```diff
@@ -2058,4 +2058,47 @@ private void awaitOrFail(CountDownLatch latch) {
...
+            try {
+                testRealm.distinct(AnnotationIndexTypes.class, "notIndex" + fieldName);
+                fail("notIndex" + fieldName);
+            } catch (UnsupportedOperationException ignore) {
+            }
+        }
+
+        try {
```
Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: Separate unit test

Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: There also seems to be missing a unit test for fields not supported like e.g. object or lists.

File: realm/src/main/java/io/realm/Realm.java (hunk 2 above)
Comment (kneth) [inline realm/src/main/java/io/realm/Realm.java]: IMHO, as a Realm doesn't have order, it doesn't really matter.

Comment (cmelchior) [inline realm/src/main/java/io/realm/Realm.java]: Lets put it this way: I should have known better and I wondered. Having a line like "As a Realm is unordered, it is undefined which element is returned for values with multiple candidates"  (could probably be rephrased)

It makes the contract of the method totally clear and might save a support ticket :)

Comment (kneth) [inline realm/src/main/java/io/realm/Realm.java]: That's a valid point.

File: realm/src/androidTest/java/io/realm/RealmTest.java [hunk 5]
```diff
@@ -2058,4 +2058,78 @@ private void awaitOrFail(CountDownLatch latch) {
...
+                obj.setIndexDate(new Date(1000 * j));
+                obj.setIndexString("Test " + j);
+                obj.setNotIndexBoolean(j % 2 == 0);
+                obj.setNotIndexLong(j);
+                obj.setNotIndexDate(new Date(1000 * j));
+                obj.setNotIndexString("Test " + j);
+            }
+        }
+        realm.commitTransaction();
+    }
```
Comment (beeender) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: One more test case for `null`? Does the `distinct` work on same rows with `null` values as well?

**Comment by cmelchior** [comment]
Ready for review @realm/java

**Comment by beeender** [comment]
:+1:

**Comment by cmelchior** [comment]
:+1:

File: realm-jni/src/io_realm_internal_table.cpp [hunk 6]
```diff
@@ -1315,17 +1315,24 @@ JNIEXPORT jlong JNICALL Java_io_realm_internal_Table_nativeGetDistinctView(
...
+        case type_Int:
+        case type_DateTime:
+        case type_String:
+            try {
+                TableView* pTableView = new TableView( pTable->get_distinct_view(S(columnIndex)) );
+                return reinterpret_cast<jlong>(pTableView);
+            } CATCH_STD()
+            break;
+        default:
+            ThrowException(env, IllegalArgument, "Invalid type - Only String, Date, boolean, short, int, long and their boxed variants are supported.");
```
Comment (nhachicha) [inline realm-jni/src/io_realm_internal_table.cpp]: in async I use this message instead 
`Distinct is not supported on binary data, object references and RealmList`

**Comment by nhachicha** [comment]
:+1:

File: realm-jni/src/io_realm_internal_table.cpp (hunk 6 above)
Comment (kneth) [inline realm-jni/src/io_realm_internal_table.cpp]: `float` and `double` ain't supported too.

File: realm/src/androidTest/java/io/realm/RealmTest.java [hunk 7]
```diff
@@ -2058,4 +2058,90 @@ private void awaitOrFail(CountDownLatch latch) {
...
+        final long numberOfBlocks = 25;
+        final long numberOfObjects = 10; // must be greater than 1
+
+        populateForDistinct(testRealm, numberOfBlocks, numberOfObjects, false);
+
+        for (String fieldName : new String[]{"Boolean", "Long", "Date", "String"}) {
+            try {
+                testRealm.distinct(AnnotationIndexTypes.class, "notIndex" + fieldName);
+                fail("notIndex" + fieldName);
+            } catch (UnsupportedOperationException ignore) {
```
Comment (bmunkholm) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: Don't we call it "expected" instead of "ignore" now?
